File: prime_test/miller_rabin/src/MillerRabinDemo.cs
// C#
// 素数判定: ミラーラビン (Miller-Rabin)

using System;

class MillerRabinDemo
{
    static Random random = new Random();
    
    static bool IsPrime(int target, int k = 40)
    {
        // 特殊なケースを処理
        if (target <= 1)
            return false;
        if (target <= 3)
            return true;
        if (target % 2 == 0)
            return false;
        
        // n-1 = 2^r * d の形に分解する（dは奇数）
        int r = 0;
        int d = target - 1;
        while (d % 2 == 0)
        {
            r++;
            d /= 2;
        }
        
        // k回の試行を実行
        for (int i = 0; i < k; i++)
        {
            int a = random.Next(2, target - 1);
            long x = ModPow(a, d, target);  // a^d mod n を計算
            
            if (x == 1 || x == target - 1)
                continue;
            
            bool isProbablyPrime = false;
            for (int j = 0; j < r - 1; j++)
            {
                x = ModPow(x, 2, target);
                if (x == target - 1)
                {
                    isProbablyPrime = true;
                    break;
                }
            }
            
            if (!isProbablyPrime)
                return false;  // 合成数と確定
        }
        
        return true;  // 素数である可能性が高い
    }
    
    // C#の BigInteger が使えない場合の ModPow の実装
    static long ModPow(long baseNum, int exponent, int modulus)
    {
        long result = 1;
        baseNum = baseNum % modulus;
        
        while (exponent > 0)
        {
            if (exponent % 2 == 1)
                result = (result * baseNum) % modulus;
            
            exponent >>= 1;
            baseNum = (baseNum * baseNum) % modulus;
        }
        
        return result;
    }
    
    static void Main()
    {
        Console.WriteLine("MillerRabin TEST -----> start");
        
        Console.WriteLine("\nIsPrime");
        int[] inputList = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 97, 100, 101, 997 };
        foreach (int input in inputList)
        {
            bool output = IsPrime(input);
            Console.WriteLine($"  {input}: {output}");
        }
        
        Console.WriteLine("\nMillerRabin TEST <----- end");
    }
}

File: prime_test/miller_rabin/src/MillerRabinDemo.java
// Java
// 素数判定: ミラーラビン (Miller-Rabin)

import java.util.Arrays;
import java.util.Random;

public class MillerRabinDemo {
    
    public static boolean isPrime(long target, int k) {
        // 特殊なケースを処理
        if (target <= 1) {
            return false;
        }
        if (target <= 3) {
            return true;
        }
        if (target % 2 == 0) {
            return false;
        }
        
        // n-1 = 2^r * d の形に分解する（dは奇数）
        int r = 0;
        long d = target - 1;
        while (d % 2 == 0) {
            r++;
            d /= 2;
        }
        
        // k回の試行を実行
        Random random = new Random();
        for (int i = 0; i < k; i++) {
            long a = 2 + random.nextInt((int)(target - 3));
            long x = modPow(a, d, target);
            
            if (x == 1 || x == target - 1) {
                continue;
            }
            
            boolean isProbablePrime = false;
            for (int j = 0; j < r - 1; j++) {
                x = modPow(x, 2, target);
                if (x == target - 1) {
                    isProbablePrime = true;
                    break;
                }
            }
            
            if (!isProbablePrime) {
                return false; // 合成数と確定
            }
        }
        
        return true; // 素数である可能性が高い
    }
    
    public static boolean isPrime(long target) {
        return isPrime(target, 40);
    }
    
    // a^b mod m を計算する関数
    private static long modPow(long base, long exponent, long modulus) {
        return java.math.BigInteger.valueOf(base)
                .modPow(java.math.BigInteger.valueOf(exponent), java.math.BigInteger.valueOf(modulus))
                .longValue();
    }
    
    public static void main(String[] args) {
        System.out.println("MillerRabin TEST -----> start");
        
        System.out.println("\nis_prime");
        long[] inputList = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 97, 100, 101, 997};
        for (long input : inputList) {
            boolean output = isPrime(input);
            System.out.println("  " + input + ": " + output);
        }
        
        System.out.println("\nMillerRabin TEST <----- end");
    }
}